{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Silent bystander edits for predicting pegRNA efficiency with PRIDICT2.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Silent bystander edits can increase the efficiency of (e)pegRNA efficiencies. This notebook allows to create inputs for PRIDICT2.0 that contain silent bystander edits, up to 5 bp up- and downstream of the edit.\n",
    "\n",
    "Requirements:\n",
    "- PRIDICT input sequence with **150 bp** context on both sides of the edit (e.g. 150bp - (A/G) - 150bp; longer context needed than the 100bp for standard PRIDICT2.0)\n",
    "- PRIDICT2.0 conda environment (includes necessary packages)\n",
    "- Replacements up to 5 bp allowed (5+ bp replacements will lead to too many options with silent bystanders to reasonably get PRIDICT2.0 predictions for all)\n",
    "\n",
    "How to use:\n",
    "- Use this notebook to create an input batch file for PRIDICT2.0 prediction with silent bystanders\n",
    "- We only provide silent bystander predictions with **1bp replacements** or **multibp replacements** (not with insertion/deletions)\n",
    "- If sequence is in exon, PRIDICT input sequence has to be *in frame* (ORF_start=0) and variable \"silent\" should be \"yes\"\n",
    "- Keep variable \"change_edit_bases\" as default \"no\" if you do not wish to change your defined edit bases even if this would lead to the same amino acid.\n",
    "  Example: NNN(GTA/CAT)NNN would be a V to H change, but same would be the case with NNN(GTA/CA**G**)NNN. If choosing \"no\" then changes within brackets will be kept and NNN(GTA/CA**G**)NNN would not be used.\n",
    "- If you want to create any bystander (also non-silent) set variable \"silent\" to \"no\".\n",
    "\n",
    "- Input: PRIDICT input format, but with 150bp flanking bases on both sides\n",
    "- Manual function: Create silent bystander PRIDICT inputs for 1 mutation/edit\n",
    "- Batch function: Get the inputs for silent bystander for all PRIDICT inputs in an input .csv file\n",
    "- Finally run PRIDICT2.0 (batch mode) with created input sequences to get efficiency predictions\n",
    "\n",
    "Optional:\n",
    "- Summarize predictions of all bystander-variants into a single file, by selecting best predicted pegRNA for each variant (last section)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import necessary packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "import re\n",
    "import pandas as pd\n",
    "from itertools import product\n",
    "import os\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions required to run notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_all_sequences(length):\n",
    "    ### Generate all possible sequences of a given length\n",
    "    nucleotides = ['A', 'T', 'G', 'C']\n",
    "    return [''.join(seq) for seq in product(nucleotides, repeat=length)]\n",
    "\n",
    "def generate_combinations(left_context_close, left_options, edit, right_options, right_context_close):\n",
    "    ### Generate all possible combinations of left and right context with edit\n",
    "    return [\n",
    "        left_context_close + left_opt + edit + right_opt + right_context_close\n",
    "        for left_opt, right_opt in product(left_options, right_options)\n",
    "    ]\n",
    "\n",
    "def convert_differences_to_lowercase(option, original):\n",
    "    ### Convert differences between option and original to lowercase\n",
    "    return ''.join(\n",
    "        opt.lower() if opt != orig else opt \n",
    "        for opt, orig in zip(option, original)\n",
    "    )\n",
    "\n",
    "def split_sequence(seq):\n",
    "    ### Split a sequence into three parts: before first lowercase letter, between first and last lowercase letter, and after last lowercase letter\n",
    "    lowercase_positions = [m.start() for m in re.finditer('[a-z]', seq)]\n",
    "    if not lowercase_positions:\n",
    "        return seq, '', '', 0\n",
    "    \n",
    "    first_lower_pos = lowercase_positions[0]\n",
    "    last_lower_pos = lowercase_positions[-1]\n",
    "    before_first_lower = seq[:first_lower_pos]\n",
    "    between_first_last_lower = seq[first_lower_pos:last_lower_pos + 1]\n",
    "    after_last_lower = seq[last_lower_pos + 1:]\n",
    "\n",
    "    return before_first_lower, between_first_last_lower, after_last_lower, len(lowercase_positions)\n",
    "\n",
    "def validate_context_length(input, minimum_flanking):\n",
    "    ### Validate that the context length is at least the minimum flanking length\n",
    "    left_context = len(input.split('(')[0])\n",
    "    right_context = len(input.split(')')[1])\n",
    "    if left_context < minimum_flanking or right_context < minimum_flanking:\n",
    "        raise ValueError('Context length is less than minimum flanking length! Please check your input sequence.')\n",
    "\n",
    "def process_contexts(input, bystander_window, close_context_len):\n",
    "    ### Process the input sequence into left and right context\n",
    "    left_context = input.split('(')[0]\n",
    "    right_context = input.split(')')[1]\n",
    "    left_context_start = left_context[:-bystander_window-close_context_len]\n",
    "    left_context_close = left_context[-bystander_window-close_context_len:-bystander_window]\n",
    "    left_bystander_window = left_context[-bystander_window:]\n",
    "    right_context_end = right_context[bystander_window+close_context_len:]\n",
    "    right_context_close = right_context[bystander_window:bystander_window+close_context_len]\n",
    "    right_bystander_window = right_context[:bystander_window]\n",
    "    return left_context_start, left_context_close, left_bystander_window, right_context_close, right_bystander_window, right_context_end\n",
    "\n",
    "def isDNA(sequence):\n",
    "    \"\"\" Check whether sequence contains only DNA bases. \"\"\"\n",
    "    onlyDNA = True\n",
    "    diff_set = set(sequence) - set('ACTGatgc')\n",
    "    if diff_set:\n",
    "        onlyDNA = False\n",
    "        print('Non-DNA bases detected. Please use ATGC.')\n",
    "        print(sequence)\n",
    "        raise ValueError\n",
    "    return onlyDNA\n",
    "\n",
    "def primesequenceparsing(sequence: str) -> object:\n",
    "    \"\"\"\n",
    "    Function which takes target sequence with desired edit as input and \n",
    "    editing characteristics as output. Edit within brackets () and original\n",
    "    equence separated by backslash from edited sequence: (A/G) == A to G mutation.\n",
    "    Placeholder for deletions and insertions is '-'.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    sequence : str\n",
    "        Target sequence with desired edit in brackets ().\n",
    "    \"\"\"\n",
    "    \n",
    "    sequence = sequence.replace('\\n','')  # remove any spaces or linebreaks in input\n",
    "    sequence = sequence.replace(' ','')\n",
    "    sequence = sequence.upper()\n",
    "    if sequence.count('(') != 1:\n",
    "        print(sequence)\n",
    "        print('More or less than one bracket found in sequence! Please check your input sequence.')\n",
    "        raise ValueError\n",
    "\n",
    "    five_prime_seq = sequence.split('(')[0]\n",
    "    three_prime_seq = sequence.split(')')[1]\n",
    "\n",
    "    sequence_set = set(sequence)\n",
    "    if '/' in sequence_set:\n",
    "        original_base = sequence.split('/')[0].split('(')[1]\n",
    "        edited_base = sequence.split('/')[1].split(')')[0]\n",
    "\n",
    "        # edit flanking bases should *not* be included in the brackets\n",
    "        if (original_base[0] == edited_base[0]) or (original_base[-1] == edited_base[-1]):\n",
    "            print(sequence)\n",
    "            print('Flanking bases should not be included in brackets! Please check your input sequence.')\n",
    "            raise ValueError\n",
    "    elif '+' in sequence_set:  #insertion\n",
    "        original_base = '-'\n",
    "        edited_base = sequence.split('+')[1].split(')')[0]\n",
    "    elif '-' in sequence_set:  #deletion\n",
    "        original_base = sequence.split('-')[1].split(')')[0]\n",
    "        edited_base = '-'\n",
    "\n",
    "    # ignore \"-\" in final sequences (deletions or insertions)\n",
    "    if original_base == '-':\n",
    "        original_seq = five_prime_seq + three_prime_seq\n",
    "        if edited_base != '-':\n",
    "            mutation_type = 'Insertion'\n",
    "            correction_length = len(edited_base)\n",
    "        else:\n",
    "            print(sequence)\n",
    "            raise ValueError\n",
    "    else:\n",
    "        original_seq = five_prime_seq + original_base + three_prime_seq\n",
    "        if edited_base == '-':\n",
    "            mutation_type = 'Deletion'\n",
    "            correction_length = len(original_base)\n",
    "        elif len(original_base) == 1 and len(edited_base) == 1:\n",
    "            if isDNA(original_base) and isDNA(edited_base):  # check if only AGCT is in bases\n",
    "                mutation_type = '1bpReplacement'\n",
    "                correction_length = len(original_base)\n",
    "            else:\n",
    "                print(sequence)\n",
    "                print('Non DNA bases found in sequence! Please check your input sequence.')\n",
    "                raise ValueError\n",
    "        elif len(original_base) > 1 or len(edited_base) > 1:\n",
    "            if isDNA(original_base) and isDNA(edited_base):  # check if only AGCT is in bases\n",
    "                mutation_type = 'MultibpReplacement'\n",
    "                if len(original_base) == len(\n",
    "                        edited_base):  # only calculate correction length if replacement does not contain insertion/deletion\n",
    "                    correction_length = len(original_base)\n",
    "                else:\n",
    "                    print(sequence)\n",
    "                    print('Only 1bp replacements or replacements of equal length (before edit/after edit) are supported! Please check your input sequence.')\n",
    "                    raise ValueError\n",
    "            else:\n",
    "                print(sequence)\n",
    "                print('Non DNA bases found in sequence! Please check your input sequence.')\n",
    "                raise ValueError\n",
    "\n",
    "    if edited_base == '-':\n",
    "        edited_seq = five_prime_seq + three_prime_seq\n",
    "    else:\n",
    "        edited_seq = five_prime_seq + edited_base.lower() + three_prime_seq\n",
    "\n",
    "    if isDNA(edited_seq) and isDNA(original_seq):  # check whether sequences only contain AGCT\n",
    "        pass\n",
    "    else:\n",
    "        raise ValueError\n",
    "\n",
    "    basebefore_temp = five_prime_seq[\n",
    "                      -1:]  # base before the edit, could be changed with baseafter_temp if Rv strand is targeted (therefore the \"temp\" attribute)\n",
    "    baseafter_temp = three_prime_seq[:1]  # base after the edit\n",
    "\n",
    "    editposition_left = len(five_prime_seq)\n",
    "    editposition_right = len(three_prime_seq)\n",
    "    return original_base, edited_base, original_seq, edited_seq, editposition_left, editposition_right, mutation_type, correction_length, basebefore_temp, baseafter_temp\n",
    "\n",
    "def bystander_creation_for_pridict(pridict_input_original, silent_surrounding_AA_nr, ORF_start, name, minimum_flanking, total_edit_limit,max_edit_length, silent='yes', change_edit_bases='no'):\n",
    "    if silent == 'no':\n",
    "        if silent_surrounding_AA_nr != 1:\n",
    "            print('*** Flanking AA number is set to 1 (maximum) for non-silent bystander, as more will lead to memory issues due to a huge amount of possibilities. ***')\n",
    "            silent_surrounding_AA_nr = 1\n",
    "\n",
    "\n",
    "\n",
    "    # Map each amino acid to its corresponding codons\n",
    "    codon_map = {\n",
    "        'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "        'C': ['TGT', 'TGC'],\n",
    "        'D': ['GAT', 'GAC'],\n",
    "        'E': ['GAA', 'GAG'],\n",
    "        'F': ['TTT', 'TTC'],\n",
    "        'G': ['GGT', 'GGC', 'GGA', 'GGG'],\n",
    "        'H': ['CAT', 'CAC'],\n",
    "        'I': ['ATT', 'ATC', 'ATA'],\n",
    "        'K': ['AAA', 'AAG'],\n",
    "        'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "        'M': ['ATG'],\n",
    "        'N': ['AAT', 'AAC'],\n",
    "        'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "        'Q': ['CAA', 'CAG'],\n",
    "        'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "        'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "        'T': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "        'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "        'W': ['TGG'],\n",
    "        'Y': ['TAT', 'TAC'],\n",
    "        '*': ['TAA', 'TAG', 'TGA']  # Stop codons\n",
    "    }\n",
    "    # assert that ORF_start is in [0, 1, 2]\n",
    "    if ORF_start not in [0, 1, 2]:\n",
    "        raise ValueError('ORF_start has to be 0, 1 or 2!')\n",
    "    \n",
    "    pridict_input_original = pridict_input_original[ORF_start:]\n",
    "    minimum_flanking = minimum_flanking - ORF_start # subtract ORF_start from minimum_flanking to account for the shift in the ORF and allow that the minimum flanking is still met\n",
    "    pridict_input_original = pridict_input_original.upper()\n",
    "    validate_context_length(pridict_input_original, minimum_flanking)\n",
    "\n",
    "    # confirm that pridict_input_original contains replacement edit:\n",
    "    mutation_type = primesequenceparsing(pridict_input_original)[6]\n",
    "    if not mutation_type in ['1bpReplacement', 'MultibpReplacement']:\n",
    "        raise ValueError('Only 1bp replacements or multi bp replacements of equal length (before edit/after edit) are supported for silent bystander predictions! Please check your input sequence.')\n",
    "\n",
    "    left_context = pridict_input_original.split('(')[0]  # bases before brackets\n",
    "    right_context = pridict_input_original.split(')')[1]  # bases after brackets\n",
    "    left_context_length = len(left_context)\n",
    "    # number of bases which are in the context, but already part of the AA that is edited\n",
    "    remaining_bases = left_context_length % 3\n",
    "\n",
    "    # sequence/base of the defined target before editing:\n",
    "    edit_before = pridict_input_original.split('(')[1].split('/')[0]\n",
    "    # sequence/base of the defined target after editing:\n",
    "    edit_after = pridict_input_original.split('/')[1].split(')')[0]\n",
    "    edit_after = convert_differences_to_lowercase(edit_after, edit_before)  # converts the actual edited bases to lowercase; not the whole edited sequence (e.g. AGA to TGT will be tGt and not tgt)\n",
    "    edit_length = len(edit_before)\n",
    "    if edit_length > max_edit_length:\n",
    "        raise ValueError(f'Only edits of up to {max_edit_length} bases are supported! Please check your input sequence.')\n",
    "\n",
    "    # sequence of the edited target including context\n",
    "    edited_sequence = left_context + edit_after + right_context\n",
    "    # sequence of the original target including context\n",
    "    original_sequence = left_context + edit_before + right_context\n",
    "\n",
    "    # number of bases which make up the AA(s) that are part of the edit (in brackets of the pridict input)\n",
    "    edit_AA_length_in_bp = math.ceil((remaining_bases + edit_length)/3)*3\n",
    "    # number of AA that are edited (silent or not)\n",
    "    edit_AA_length = int(edit_AA_length_in_bp / 3)\n",
    "    downstream_AA_context_basenr = edit_AA_length_in_bp - edit_length - remaining_bases\n",
    "\n",
    "    # sequence upstream that is unchanged (also not silently changed)\n",
    "    untouched_left_context = left_context[:-remaining_bases-silent_surrounding_AA_nr*3]\n",
    "    # sequence downstream that is unchanged (also not silently changed)\n",
    "    untouched_right_context = right_context[downstream_AA_context_basenr+silent_surrounding_AA_nr*3:]\n",
    "\n",
    "    # sequence of bases that can be changed by the silent bystander mutations (including edit bases)\n",
    "    start_pos = len(left_context)-remaining_bases-silent_surrounding_AA_nr*3\n",
    "    end_pos = len(left_context)-remaining_bases+edit_AA_length_in_bp+silent_surrounding_AA_nr*3\n",
    "\n",
    "    # print(end_pos, edit_AA_length_in_bp, downstream_AA_context_basenr, silent_surrounding_AA_nr*3, remaining_bases)\n",
    "\n",
    "    original_bases = original_sequence[start_pos:end_pos]\n",
    "    original_AA_seq = str(Seq(original_bases).translate())\n",
    "\n",
    "    potentially_changed_bases = edited_sequence[start_pos:end_pos]\n",
    "    potentially_changed_AA = str(Seq(potentially_changed_bases).translate())\n",
    "    # print(potentially_changed_bases)\n",
    "    if silent == 'yes':\n",
    "    # create all possible codon variants which result in the same AA sequence\n",
    "        bystander_option_list = []\n",
    "        for aa in potentially_changed_AA:\n",
    "            amino_acid_codons = codon_map[aa]\n",
    "            bystander_option_list.append(amino_acid_codons)\n",
    "\n",
    "        # generate all possible combinations of codons in the bystander_option_list. For each AA, we have a list of possible codons in there\n",
    "        all_combinations = list(product(*bystander_option_list))\n",
    "        # print(len(all_combinations))\n",
    "        result_strings = [''.join(combination) for combination in all_combinations]\n",
    "\n",
    "        # remove the mutated sequence without any bystander from this list\n",
    "        result_strings = [result_string for result_string in result_strings if result_string.upper() != potentially_changed_bases.upper()]\n",
    "        print('Number of silent options INCLUDING changed edit bases', len(result_strings))\n",
    "\n",
    "        if change_edit_bases == 'no':\n",
    "            # remove options where also defined edited bases are changed (silently); check whether every position where we have a lowercase letter in potentially_changed_bases. this position should be the same base in the result_string and potentially_changed_bases\n",
    "            result_strings = [result_string for result_string in result_strings if all([result_string[i].upper() == potentially_changed_bases[i].upper() for i in range(len(result_string)) if potentially_changed_bases[i].islower()])]\n",
    "            print('Number of silent options EXCLUDING changed edit bases', len(result_strings))\n",
    "\n",
    "        result_AAs = [str(Seq(result_string).translate()) for result_string in result_strings]\n",
    "\n",
    "\n",
    "    elif silent == 'no':\n",
    "        result_strings = generate_all_sequences(len(potentially_changed_bases))\n",
    "        # print('generating sequences done')\n",
    "\n",
    "\n",
    "        # remove the mutated sequence without any bystander from this list\n",
    "        result_strings = [result_string for result_string in result_strings if result_string.upper() != potentially_changed_bases.upper()]\n",
    "        print('Number of silent options INCLUDING changed edit bases', len(result_strings))\n",
    "\n",
    "        if change_edit_bases == 'no':\n",
    "            # remove options where also defined edited bases are changed (silently); check whether every position where we have a lowercase letter in potentially_changed_bases. this position should be the same base in the result_string and potentially_changed_bases\n",
    "            result_strings = [result_string for result_string in result_strings if all([result_string[i].upper() == potentially_changed_bases[i].upper() for i in range(len(result_string)) if potentially_changed_bases[i].islower()])]\n",
    "            print('Number of silent options EXCLUDING changed edit bases', len(result_strings))\n",
    "\n",
    "        print('before')\n",
    "        result_AAs = [str(Seq(result_string).translate()) for result_string in result_strings]\n",
    "        print('after')\n",
    "\n",
    "    # for each element of result_string, check whether a base is different from the original sequence. If so, convert it to lowercase\n",
    "    result_strings = [convert_differences_to_lowercase(result_string, original_bases) for result_string in result_strings]\n",
    "\n",
    "    full_edited_list = [untouched_left_context + result_string + untouched_right_context for result_string in result_strings]\n",
    "\n",
    "    pridict_input_lst = []\n",
    "    totalbasechanges_lst = []\n",
    "    edit_name_bystander_lst = []\n",
    "    original_length_lst = []\n",
    "    final_length_lst = []\n",
    "    edited_bystander_focus_sequence_lst = []\n",
    "    edited_bystander_focus_AA_lst = []\n",
    "    editonly_focus_sequence_lst = []\n",
    "    editonly_nobystander_focus_AA_lst = []\n",
    "    for index, sequence in enumerate(full_edited_list):\n",
    "        unchanged_before, edit, unchanged_after, totalbasechanges = split_sequence(sequence)\n",
    "        original_base = original_sequence[len(unchanged_before):-len(unchanged_after)]\n",
    "        final_pridict_seq = f\"{unchanged_before}({original_base}/{edit}){unchanged_after}\"\n",
    "        edit_name_bystander = f\"{name}_{original_base}_{edit}\"\n",
    "        edited_bystander_focuse_sequence = result_strings[index]\n",
    "        edited_bystander_focus_AA = result_AAs[index]\n",
    "        edit_name_bystander_lst.append(edit_name_bystander)\n",
    "        final_length_lst.append(len(edit))\n",
    "        pridict_input_lst.append(final_pridict_seq)\n",
    "        totalbasechanges_lst.append(totalbasechanges)\n",
    "        original_length_lst.append(edit_length)\n",
    "        edited_bystander_focus_sequence_lst.append(edited_bystander_focuse_sequence)\n",
    "        edited_bystander_focus_AA_lst.append(edited_bystander_focus_AA)\n",
    "        editonly_focus_sequence_lst.append(potentially_changed_bases)\n",
    "        editonly_nobystander_focus_AA_lst.append(potentially_changed_AA)\n",
    "  \n",
    "    print(f'Number of possible bystander mutations: {len(result_strings)}')\n",
    "    print('Sequence flanking edit position (before edit):',original_bases)\n",
    "    print('Sequence flanking edit position (AFTER edit, without bystander editing):',potentially_changed_bases)\n",
    "    print('AA flanking edit position (before edit):',original_AA_seq)\n",
    "    print('AA flanking edit position (after edit, without bystander editing):',potentially_changed_AA)\n",
    "    print('AA flanking edit position (after edit, WITH bystander editing:',set(result_AAs))\n",
    "\n",
    "    finaldf = pd.DataFrame(list(zip(\n",
    "        edit_name_bystander_lst, pridict_input_lst, original_length_lst, final_length_lst, totalbasechanges_lst, edited_bystander_focus_sequence_lst, edited_bystander_focus_AA_lst, editonly_focus_sequence_lst, editonly_nobystander_focus_AA_lst\n",
    "    )), columns=['sequence_name', 'editseq', 'original_edit_length', 'final_edit_length_with_bystander', 'total_nr_of_base_changes', 'bystander_focus_sequence', 'bystander_focus_AA', 'editedonly_focus_sequence', 'editedonly_focus_AA'])\n",
    "\n",
    "    # filter finaldf to exclude rows where final_edit_length_with_bystander > total_edit_limit\n",
    "    finaldf = finaldf[finaldf.final_edit_length_with_bystander <= total_edit_limit]\n",
    "    return finaldf\n",
    "\n",
    "def bystander_input_generator(inputpath, inputfilename, outputpath, outputfilename, silent_surrounding_AA_nr, total_edit_limit, max_edit_length, minimum_flanking):\n",
    "    df = pd.read_csv(inputpath+inputfilename)\n",
    "    all_dfs = []\n",
    "    for _, row in df.iterrows():\n",
    "        name = row[\"Name\"]\n",
    "        silent = row[\"silent\"]\n",
    "        change_edit_bases = row[\"change_edit_bases\"]\n",
    "        print(name)\n",
    "        ORF_start = row[\"in_frame\"]\n",
    "        if ORF_start == 'yes':\n",
    "            ORF_start = 0\n",
    "        else:\n",
    "            raise ValueError('ORF start is not in frame! Please check your input sequence. Input sequence has to be in-frame.')\n",
    "        pridict_input_original = row.pridict_input\n",
    "        silentbystanderdf = bystander_creation_for_pridict(pridict_input_original, silent_surrounding_AA_nr, ORF_start, name, minimum_flanking, total_edit_limit,max_edit_length, silent=silent, change_edit_bases=change_edit_bases)\n",
    "        print(f'{len(silentbystanderdf)} silent bystander sequences created for {name}')\n",
    "        print()\n",
    "        all_dfs.append(silentbystanderdf)\n",
    "    final_df = pd.concat(all_dfs, ignore_index=True)\n",
    "    final_df.to_csv(outputpath+outputfilename)\n",
    "    return df, final_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Manual mode:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of silent options INCLUDING changed edit bases 191\n",
      "Number of silent options EXCLUDING changed edit bases 191\n",
      "Number of possible bystander mutations: 191\n",
      "Sequence flanking edit position (before edit): GATTCTGATGAGCCT\n",
      "Sequence flanking edit position (AFTER edit, without bystander editing): GATTCTcATGAGCCT\n",
      "AA flanking edit position (before edit): DSDEP\n",
      "AA flanking edit position (after edit, without bystander editing): DSHEP\n",
      "AA flanking edit position (after edit, WITH bystander editing: {'DSHEP'}\n"
     ]
    }
   ],
   "source": [
    "### Required variables (adapt to your needs)\n",
    "name = 'test1_cftr'\n",
    "pridict_input_original = 'TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTGGGGAAAAAAGGAAGAATTCTATTCTCAATCCAATCAACTCTATACGAAAATTTTCCATTGTGCAAAAGACTCCCTTACAAATGAATGGCATCGAAGAGGATTCT(G/C)ATGAGCCTTTAGAGAGAAGGCTGTCCTTAGTACCAGATTCTGAGCAGGGAGAGGCGATACTGCCTCGCATCAGCGTGATCAGCACTGGCCCCACGCTTCAGGCACGAAGGAGGCAGTCTGTCCTGAACCTGATGACACACTCAGTTAACC'\n",
    "silent = 'yes'  # default: yes; put to 'no' if you want to get all possible bystander mutations, including non-silent mutations which change the AA sequence\n",
    "change_edit_bases='no'  # default: no; put to 'yes' if you want to get all possible silent bystander mutations, including those which also change the edit bases you defined in the input (if applicable)\n",
    "###\n",
    "\n",
    "### default variables (do not change unless you want to adapt the script)\n",
    "silent_surrounding_AA_nr = 2 # number of amino acids up- and downstream of edit AA for which silent bystander mutations will be created; default = 2; maximum 4 AA (up- and downstream) are allowed in this script\n",
    "ORF_start = 0  # 0 means that the ORF starts at the first base of the input sequence; if the ORF starts at the second or third base, change this to 1 or 2, respectively\n",
    "total_edit_limit = 40 # limit maximum length of edit (including bystander edits) to 40 bases; max. edit length for PRIDICT2.0 predictions is 40 bases\n",
    "max_edit_length = 10 # maximum length of the edit bases you want to change\n",
    "minimum_flanking = 94  # minimum edit-flanking length, after correcting for ORF_start. Only required as sanity check; do not change.\n",
    "###\n",
    "\n",
    "# run manual bystander creation:\n",
    "silentbystanderdf = bystander_creation_for_pridict(pridict_input_original, silent_surrounding_AA_nr, ORF_start, name, minimum_flanking, total_edit_limit,max_edit_length, silent=silent, change_edit_bases=change_edit_bases)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>editseq</th>\n",
       "      <th>original_edit_length</th>\n",
       "      <th>final_edit_length_with_bystander</th>\n",
       "      <th>total_nr_of_base_changes</th>\n",
       "      <th>bystander_focus_sequence</th>\n",
       "      <th>bystander_focus_AA</th>\n",
       "      <th>editedonly_focus_sequence</th>\n",
       "      <th>editedonly_focus_AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>test1_cftr_TCTGATGAG_aaacAaaAa</td>\n",
       "      <td>TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>aaacAaaAa</td>\n",
       "      <td>KQK</td>\n",
       "      <td>TCTcATGAG</td>\n",
       "      <td>SHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>test1_cftr_TCTGATGAG_aaacAaaAt</td>\n",
       "      <td>TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>aaacAaaAt</td>\n",
       "      <td>KQN</td>\n",
       "      <td>TCTcATGAG</td>\n",
       "      <td>SHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>test1_cftr_TCTGATG_aaacAaa</td>\n",
       "      <td>TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>aaacAaaAG</td>\n",
       "      <td>KQK</td>\n",
       "      <td>TCTcATGAG</td>\n",
       "      <td>SHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>test1_cftr_TCTGATGAG_aaacAaaAc</td>\n",
       "      <td>TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>aaacAaaAc</td>\n",
       "      <td>KQN</td>\n",
       "      <td>TCTcATGAG</td>\n",
       "      <td>SHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>test1_cftr_TCTGATGAG_aaacAaata</td>\n",
       "      <td>TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>aaacAaata</td>\n",
       "      <td>KQI</td>\n",
       "      <td>TCTcATGAG</td>\n",
       "      <td>SHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>test1_cftr_TCTGATGAG_aaacAaatt</td>\n",
       "      <td>TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>aaacAaatt</td>\n",
       "      <td>KQI</td>\n",
       "      <td>TCTcATGAG</td>\n",
       "      <td>SHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>test1_cftr_TCTGATGA_aaacAaat</td>\n",
       "      <td>TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>aaacAaatG</td>\n",
       "      <td>KQM</td>\n",
       "      <td>TCTcATGAG</td>\n",
       "      <td>SHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>test1_cftr_TCTGATGAG_aaacAaatc</td>\n",
       "      <td>TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>aaacAaatc</td>\n",
       "      <td>KQI</td>\n",
       "      <td>TCTcATGAG</td>\n",
       "      <td>SHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>test1_cftr_TCTGATGAG_aaacAaaga</td>\n",
       "      <td>TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>aaacAaaga</td>\n",
       "      <td>KQR</td>\n",
       "      <td>TCTcATGAG</td>\n",
       "      <td>SHE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>test1_cftr_TCTGATGAG_aaacAaagt</td>\n",
       "      <td>TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>aaacAaagt</td>\n",
       "      <td>KQS</td>\n",
       "      <td>TCTcATGAG</td>\n",
       "      <td>SHE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    sequence_name  \\\n",
       "0  test1_cftr_TCTGATGAG_aaacAaaAa   \n",
       "1  test1_cftr_TCTGATGAG_aaacAaaAt   \n",
       "2      test1_cftr_TCTGATG_aaacAaa   \n",
       "3  test1_cftr_TCTGATGAG_aaacAaaAc   \n",
       "4  test1_cftr_TCTGATGAG_aaacAaata   \n",
       "5  test1_cftr_TCTGATGAG_aaacAaatt   \n",
       "6    test1_cftr_TCTGATGA_aaacAaat   \n",
       "7  test1_cftr_TCTGATGAG_aaacAaatc   \n",
       "8  test1_cftr_TCTGATGAG_aaacAaaga   \n",
       "9  test1_cftr_TCTGATGAG_aaacAaagt   \n",
       "\n",
       "                                             editseq  original_edit_length  \\\n",
       "0  TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...                     1   \n",
       "1  TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...                     1   \n",
       "2  TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...                     1   \n",
       "3  TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...                     1   \n",
       "4  TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...                     1   \n",
       "5  TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...                     1   \n",
       "6  TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...                     1   \n",
       "7  TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...                     1   \n",
       "8  TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...                     1   \n",
       "9  TGGACAGAAACAAAAAAACAATCTTTTAAACAGACTGGAGAGTTTG...                     1   \n",
       "\n",
       "   final_edit_length_with_bystander  total_nr_of_base_changes  \\\n",
       "0                                 9                         7   \n",
       "1                                 9                         7   \n",
       "2                                 7                         6   \n",
       "3                                 9                         7   \n",
       "4                                 9                         8   \n",
       "5                                 9                         8   \n",
       "6                                 8                         7   \n",
       "7                                 9                         8   \n",
       "8                                 9                         8   \n",
       "9                                 9                         8   \n",
       "\n",
       "  bystander_focus_sequence bystander_focus_AA editedonly_focus_sequence  \\\n",
       "0                aaacAaaAa                KQK                 TCTcATGAG   \n",
       "1                aaacAaaAt                KQN                 TCTcATGAG   \n",
       "2                aaacAaaAG                KQK                 TCTcATGAG   \n",
       "3                aaacAaaAc                KQN                 TCTcATGAG   \n",
       "4                aaacAaata                KQI                 TCTcATGAG   \n",
       "5                aaacAaatt                KQI                 TCTcATGAG   \n",
       "6                aaacAaatG                KQM                 TCTcATGAG   \n",
       "7                aaacAaatc                KQI                 TCTcATGAG   \n",
       "8                aaacAaaga                KQR                 TCTcATGAG   \n",
       "9                aaacAaagt                KQS                 TCTcATGAG   \n",
       "\n",
       "  editedonly_focus_AA  \n",
       "0                 SHE  \n",
       "1                 SHE  \n",
       "2                 SHE  \n",
       "3                 SHE  \n",
       "4                 SHE  \n",
       "5                 SHE  \n",
       "6                 SHE  \n",
       "7                 SHE  \n",
       "8                 SHE  \n",
       "9                 SHE  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview of silentbystanderdf:\n",
    "silentbystanderdf.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Batch mode:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test_mutation_1_cftr\n",
      "Number of silent options INCLUDING changed edit bases 191\n",
      "Number of silent options EXCLUDING changed edit bases 191\n",
      "Number of possible bystander mutations: 191\n",
      "Sequence flanking edit position (before edit): GATTCTGATGAGCCT\n",
      "Sequence flanking edit position (AFTER edit, without bystander editing): GATTCTcATGAGCCT\n",
      "AA flanking edit position (before edit): DSDEP\n",
      "AA flanking edit position (after edit, without bystander editing): DSHEP\n",
      "AA flanking edit position (after edit, WITH bystander editing: {'DSHEP'}\n",
      "191 silent bystander sequences created for test_mutation_1_cftr\n",
      "\n",
      "test_mutation_2_brca1\n",
      "Number of silent options INCLUDING changed edit bases 31\n",
      "Number of silent options EXCLUDING changed edit bases 31\n",
      "Number of possible bystander mutations: 31\n",
      "Sequence flanking edit position (before edit): AAGGAAGAAAATCAA\n",
      "Sequence flanking edit position (AFTER edit, without bystander editing): AAGGAAaAAAATCAA\n",
      "AA flanking edit position (before edit): KEENQ\n",
      "AA flanking edit position (after edit, without bystander editing): KEKNQ\n",
      "AA flanking edit position (after edit, WITH bystander editing: {'KEKNQ'}\n",
      "31 silent bystander sequences created for test_mutation_2_brca1\n",
      "\n",
      "test_mutation_3_diaph3\n",
      "Number of silent options INCLUDING changed edit bases 383\n",
      "Number of silent options EXCLUDING changed edit bases 383\n",
      "Number of possible bystander mutations: 383\n",
      "Sequence flanking edit position (before edit): CCAGGAATGCGGATG\n",
      "Sequence flanking edit position (AFTER edit, without bystander editing): CCAGGAgTGCGGATG\n",
      "AA flanking edit position (before edit): PGMRM\n",
      "AA flanking edit position (after edit, without bystander editing): PGVRM\n",
      "AA flanking edit position (after edit, WITH bystander editing: {'PGVRM'}\n",
      "383 silent bystander sequences created for test_mutation_3_diaph3\n",
      "\n",
      "test_mutation_1_cftr_alternative\n",
      "*** Flanking AA number is set to 1 (maximum) for non-silent bystander, as more will lead to memory issues due to a huge amount of possibilities. ***\n",
      "Number of silent options INCLUDING changed edit bases 262143\n",
      "Number of silent options EXCLUDING changed edit bases 16383\n",
      "before\n",
      "after\n",
      "Number of possible bystander mutations: 16383\n",
      "Sequence flanking edit position (before edit): TCTGATGAG\n",
      "Sequence flanking edit position (AFTER edit, without bystander editing): TCTcAcGAG\n",
      "AA flanking edit position (before edit): SDE\n",
      "AA flanking edit position (after edit, without bystander editing): SHE\n",
      "AA flanking edit position (after edit, WITH bystander editing: {'TRQ', '*HH', 'VHY', 'FRN', 'LHF', 'EPV', 'KHL', 'RRD', 'CPW', 'MLR', 'NPP', 'MLF', 'IRP', 'FRI', 'LP*', 'QLT', '*HA', 'FRS', '*HR', 'GLG', 'WLI', 'HLT', 'RRS', 'YHY', 'MPG', 'EPI', 'GHS', 'FPQ', 'QLN', 'KRG', 'DHH', 'HLG', 'TPP', 'ILF', 'DLQ', 'LPH', 'DLT', 'TPI', 'QRS', 'NRF', 'NHQ', 'VPW', 'CR*', 'LRM', 'VRH', 'FLC', 'ERE', 'SHK', 'DLH', '*PT', 'WLF', 'WLM', 'MLL', 'WHF', 'VL*', 'PHC', 'WPV', 'KPR', 'CLM', 'GPW', 'RPD', 'YRA', 'VHQ', 'HPC', 'KPD', 'HPM', 'MHG', 'SHR', 'TPA', 'GRL', 'YRN', 'CRD', 'FPG', 'YRQ', 'NLC', 'LLW', 'GPR', 'TLE', 'TPL', 'NPN', 'TPK', 'LRP', '*LT', 'YHV', 'FRE', 'LRE', 'LRF', 'VPA', 'GPM', 'CPF', 'WLT', 'FLW', 'CRH', 'IRW', 'MP*', 'THI', 'DRW', 'ERH', 'PRW', 'NHN', 'IHA', 'FH*', 'ALD', 'QPM', 'WHE', 'SPC', 'GPC', 'NRL', 'SHM', 'DLL', 'CHD', 'NHH', 'GLP', 'FHY', 'SLT', 'ERQ', 'MHT', 'GPP', 'PRA', 'KRH', 'KPF', 'WHA', 'ARE', 'KLV', 'DRF', 'IHS', 'KLE', 'QHQ', 'VHA', 'HRR', 'QHM', 'IRM', 'FHI', 'RHG', 'VHD', 'PLV', 'DLW', 'PPY', 'APT', 'RPG', 'ALS', 'WRE', 'ERW', 'QRV', 'AHL', 'CRV', 'THY', 'DHD', 'HPV', 'IPN', 'VRM', 'QHT', 'NHD', 'EP*', 'EPP', 'KHA', 'LRK', 'SPV', 'WPN', 'CRG', 'WPG', 'IPH', 'CLP', '*LW', 'EHF', 'FHQ', 'VRP', 'QPE', 'PRH', 'DLR', 'WRW', '*LM', 'PLY', 'FPD', '*HI', 'ALE', 'SPD', 'THH', 'QPD', 'HRH', 'QRN', 'NLD', 'HRW', 'LLS', 'TRF', 'ARR', 'EHP', 'WLG', 'GHP', 'KHP', 'DHV', 'APW', 'CLT', 'EPN', 'ELF', 'PHS', 'RPK', 'VHC', 'SRV', 'DHA', 'QRE', 'GL*', 'NL*', 'YHD', 'CLR', 'PRM', 'CLQ', 'KHQ', 'NLE', 'EPH', 'RRG', 'NLQ', 'FLQ', 'MHN', 'RHW', 'IPA', 'CHW', 'CRY', 'SPT', 'VPV', 'SLC', 'SPP', 'IHH', '*RH', 'KLH', 'QRY', 'WPP', 'CRN', 'YHE', 'MLA', 'IPL', 'YPA', 'WLH', 'DRV', 'HRI', 'LPE', 'LLL', 'QHK', 'YPE', 'ARK', 'HL*', 'LHK', 'LL*', 'HHA', 'MPL', 'NHG', 'HLA', 'AHP', '*LR', 'QPP', 'DHY', 'PLW', '*RI', 'NLN', 'YPH', 'NLK', 'SHA', '*HQ', 'YLY', '*LC', 'RLT', '*RE', 'ER*', 'MLY', 'GHN', 'IRT', 'CLY', 'LPY', 'GHT', 'WPA', 'IRQ', 'FLE', 'RPF', 'QRF', 'ELK', 'ARG', 'DRE', 'FPR', 'YHT', 'CPT', 'YRE', '*LP', 'WHI', 'QRM', 'TLI', 'QRR', 'DHE', 'NLS', 'NLL', 'FHF', 'PPI', 'FHN', 'YRC', 'YHR', 'FHE', 'NRQ', 'YHM', 'VHT', 'KHC', 'ERC', 'APS', 'ELG', 'GHF', 'EPE', 'QPI', 'PLF', 'LPA', 'QHH', '*PG', 'YRK', 'LPR', 'PLM', 'WHC', 'ARW', 'VHG', 'KLK', 'DHQ', 'RHC', 'QHA', 'KRE', '*PP', 'QPA', 'LPM', 'QPY', 'KHG', 'ALI', 'PRK', 'MLW', '*HC', 'VLL', 'IPQ', 'KPH', 'FPK', 'GLQ', 'EPS', '*R*', 'HPF', 'FLD', 'WRG', 'QHP', 'WPD', 'MRE', 'RH*', 'GPK', 'FHM', 'RLF', 'FPL', 'IHY', 'YLK', 'HPW', 'GLD', 'ALC', 'APY', 'RPS', 'MHR', 'YLD', 'HRV', 'LPQ', 'LPP', 'GPS', 'DPN', 'NRT', 'MHW', 'FRQ', '*LI', 'SLI', 'HRT', 'MRI', '*PH', 'YRY', 'QLQ', 'HPE', 'WHM', 'WRV', 'SLW', 'FHC', 'THK', 'AHM', '*HK', 'LLR', 'RHY', '*HS', 'CLV', 'QLD', 'KR*', 'THL', 'TLP', 'NHS', 'ERS', 'RLH', 'DLN', 'APN', 'PLI', 'YPC', 'CLD', 'PRI', 'NPA', 'LRH', 'IL*', 'NLF', '*RQ', 'QRH', 'SHD', 'MRF', 'WRH', 'CPA', 'KHF', 'VRR', 'FLY', 'ALP', 'QHL', 'KPC', 'FP*', 'DLK', 'MPF', 'HLE', 'ERK', 'MPV', 'GRF', 'EHQ', 'MLK', 'ALL', 'VRQ', 'EHS', 'FLF', 'FLM', 'FHR', 'DHN', 'NHA', 'RLE', 'WLE', 'CHQ', 'GHK', 'RLQ', 'NLP', 'NR*', 'APG', 'QPV', 'LRG', 'IRR', 'EHR', 'FPA', 'TRL', 'LHI', 'HRE', 'NPC', 'CHN', 'ELH', 'TLS', 'HLW', 'PRV', 'VRK', '*RP', 'KP*', 'PRN', 'WPC', 'LRQ', 'KLN', 'MHH', 'MRP', 'KPT', 'NHE', 'NPW', 'NLH', 'NRI', 'DLV', 'MRA', 'LHP', 'FRA', 'NPS', 'LRS', 'GRH', 'NPM', 'ARA', 'IHM', 'VRY', 'KPA', 'SHW', 'RHI', 'CHF', 'AHN', 'YLC', 'IRN', 'KLW', 'WHP', 'WRF', 'VLT', 'DRC', 'CLK', 'PLC', 'MHQ', 'CPI', 'WH*', 'NHF', 'LHA', 'PLN', 'GPN', 'VLV', '*HF', 'FLR', 'CLF', 'WRA', 'WRP', 'PPN', 'LLG', 'VPI', 'GRP', 'WPF', 'YPM', 'TPD', 'MRY', 'VRW', 'QRP', 'MHI', 'CHG', 'CRL', 'ART', 'SRR', 'IHG', 'SPA', 'AHA', 'IPY', 'PPM', 'GRD', 'RR*', 'MHC', 'AHQ', 'SPW', 'PLG', 'YLI', 'HRL', 'DHI', 'AHR', 'CPY', 'CHS', 'LHN', 'IPI', 'KLQ', 'ELT', 'YRM', 'DRH', 'THG', 'SHC', 'TPM', 'WRY', 'CLL', 'QH*', 'NLT', 'SLL', 'MLN', 'SRT', 'CHC', 'DPM', 'FLK', 'CHT', 'AHS', 'SRI', 'TLW', 'IRI', '*HL', 'WHW', 'GHM', 'GRS', 'PRF', 'WLA', 'WRN', 'CLC', 'FHP', 'PPR', 'QHS', 'KPV', 'MPA', 'ERT', 'VRT', 'ARC', 'PLK', '*RS', 'LLI', 'NPF', 'EPF', 'VHM', '*RN', 'PHN', 'YRV', 'DLG', 'FPE', 'WRT', 'DPQ', 'KHW', 'NRR', 'ERF', 'FHS', '*RD', 'EHM', 'CPL', 'KLL', 'HHR', 'SRQ', 'RRY', '*PN', 'THW', 'MRG', 'DLM', 'DPT', 'TRV', 'SLG', '*PL', 'WLS', 'SPQ', 'GPI', 'EPW', 'GLA', 'SLV', 'CRW', 'DRQ', 'NLA', 'GRI', 'HRF', 'QRT', 'PRQ', 'ARQ', 'QLE', 'APL', 'DPA', 'FHH', 'FRF', 'HLM', 'VPR', 'MR*', 'GPD', 'GPA', 'ELW', 'GHV', 'TL*', 'QLF', 'HLS', 'EPQ', 'ELN', 'RRN', 'MPT', 'ERG', 'ALH', 'WRL', '*HV', 'CRK', 'GRT', 'GH*', 'RLD', 'ALW', '*PE', 'QRK', 'PRC', 'FLI', 'YHA', 'MPD', 'IHP', 'TLG', 'ILP', 'CL*', 'HRA', 'WRR', 'KLR', 'FRW', 'WRC', 'VPL', 'SRC', 'VLF', 'VHW', 'VRA', 'MPC', 'RHV', 'LLC', 'KHK', 'VLQ', 'YRH', 'IRE', 'CH*', 'LHW', 'CHI', 'LR*', 'DPH', 'QLH', 'HLL', 'QRD', 'EPD', 'DPG', 'RRI', 'RLV', 'SPR', 'VLS', 'WLY', 'EL*', 'PLT', 'TLL', 'LLD', 'ALF', 'QLW', 'HRY', 'ALQ', 'SHE', 'GHC', 'DLS', 'HLP', 'HHV', 'ELP', 'ILD', 'RPW', 'HHE', 'ERL', 'DPY', 'IP*', 'EHW', 'GHW', 'LRV', 'PRR', '*RG', 'AHY', 'QPN', 'YRW', 'RLC', 'MHA', 'TRN', 'DPW', 'SHS', 'WHH', 'HLF', 'WPT', 'VHI', 'TLF', 'VHL', 'PRP', 'NRW', 'WHD', 'VRC', 'PR*', 'KHV', 'WHY', 'AHT', 'WLK', 'THV', 'WP*', 'VLK', 'VPN', 'THD', 'HPY', 'LRA', 'IHQ', 'RPP', 'EPR', '*LS', '*H*', 'NLY', 'HHG', 'VLE', 'MPM', 'RLR', 'VPD', 'FRH', 'WLR', 'PHP', 'IRF', 'HHM', '*HM', 'MLC', 'DR*', 'FRR', 'PPL', 'QHN', 'RHL', 'ALN', 'PPD', 'RRH', 'SRA', 'WPI', 'IPM', 'HPN', 'FL*', 'VRI', 'GRC', 'TLT', 'TP*', 'FLT', 'QPF', 'CLN', 'MPS', 'TPY', 'PRL', 'NRG', 'IPW', 'MRV', 'RPI', 'ARY', 'GHY', 'CHM', 'IPR', 'WPW', 'YPY', 'FRV', 'ILV', 'PRE', 'GHQ', 'VPQ', 'NHC', 'MPR', 'CLE', 'LRY', 'TLY', 'FPS', 'IHK', 'RPN', 'LLV', '*RW', 'PLH', 'DRD', 'GHI', 'LPG', 'FPN', 'CLA', 'RPM', '*RK', 'CRF', 'VPS', 'FHW', 'LLM', 'PHM', 'EHH', 'YR*', 'AHV', 'MH*', 'HP*', 'NRV', 'CPP', 'FHK', 'YLF', 'EHY', 'SPE', 'VLY', 'WLV', 'YRI', 'YLA', 'THA', 'TRD', 'QLL', 'EHA', 'GHD', 'YLS', 'IHL', 'YRL', '*LF', 'KLI', 'QHV', 'PPV', 'SRS', 'NPR', 'PPP', 'CHL', 'QPC', 'ELE', 'YLW', 'WLC', 'NLW', 'LPT', 'IPC', 'TLA', 'VPK', 'CPQ', 'KRR', 'IHT', 'YPF', 'ILW', '*LL', 'ILM', 'SLS', 'PHE', 'PHL', 'SHI', 'SLQ', 'LPI', 'GRN', 'RHN', 'ILR', 'ARH', 'NPT', 'IPP', 'NLI', 'RRV', 'NRC', 'LRI', 'CLW', 'MLD', 'FHV', 'SPS', 'VRG', 'ELS', 'KPY', 'SPK', 'VHN', 'HHI', 'QR*', 'NPG', 'WL*', 'PHH', 'APC', 'KL*', 'YPV', 'KHE', 'RPQ', 'AR*', 'SHG', 'HHC', 'DRG', 'YHP', 'TRH', 'VRE', 'AH*', 'IPV', '*HN', 'MRC', 'PHV', 'CPH', 'PLD', 'CPV', 'FPI', 'FHG', 'MLS', '*LA', 'TLD', 'GRV', 'QPH', 'RRL', 'YLR', '*LV', 'ARS', 'QLV', 'PLA', 'VHS', 'GPG', 'PPA', 'GRQ', 'TRM', 'DHM', 'CHA', 'EHE', 'TRY', 'QLI', 'THT', 'FLN', 'DLF', 'DRR', 'NPK', 'IRA', 'YLE', 'HLR', 'EHT', 'HHL', 'CRE', 'THP', 'IRY', 'LHC', 'KRY', 'QHI', '*PF', 'GLT', 'PLL', 'QLP', 'PHR', 'VLD', 'GLK', 'IHR', 'PPC', 'DLE', 'KLM', 'CHK', 'SHN', 'CRA', 'IRD', 'FPT', 'YPL', 'TLQ', 'VPP', 'FPH', 'DPK', 'NRN', 'FLG', 'HLI', 'ALK', 'NRH', 'YPK', 'EHI', 'DHG', 'SLY', 'MLG', 'WHL', 'HRM', 'KPK', 'KHS', 'MHF', 'GRR', 'TPE', 'SHT', 'FLH', 'DPV', 'KLA', '*HG', 'SRD', 'KPL', 'SPG', 'FR*', 'TLK', 'CHR', 'EHC', 'QPQ', 'QPK', 'ILT', 'CRP', 'IPS', 'LPW', 'ELC', '*HY', 'TPC', 'WHN', 'APD', 'RPY', 'YLT', 'RHS', 'SLP', '*HP', 'APP', 'RLA', 'HHS', '*PM', 'YHH', 'ILC', 'VHV', 'WRD', 'GRK', 'QPG', 'GLY', 'AHE', 'MLQ', 'YLN', 'TH*', 'HLK', 'HPP', 'YRT', 'NRD', 'HHN', '*RY', 'QHR', 'DRN', 'HHY', 'NRM', 'MLI', 'WLD', 'LHG', 'ERA', 'YHK', 'ARF', 'DHW', 'YLG', 'VPE', 'YPW', 'ALT', 'VHK', 'VLC', 'GRE', 'LPC', 'DHC', 'PP*', 'CPR', 'WLP', 'MRD', 'VHR', 'VLR', '*L*', 'SR*', 'PHW', 'THN', 'TPS', 'MRR', 'PHF', 'QLG', 'LRR', 'YHS', 'SH*', 'YHF', 'CRQ', 'ARD', 'LHL', 'FLS', 'WRQ', 'IPF', 'MHD', 'RPT', 'TRK', 'AHK', 'HR*', 'SHV', 'SRF', 'TLH', 'THR', 'APM', 'HRG', 'WHK', 'WRS', 'GPY', 'HLQ', 'WPE', 'QHW', 'FHD', 'PHQ', '*PD', 'KRP', 'WPL', 'QL*', 'TRG', 'MLP', 'MRQ', 'LRN', 'GHG', 'TPQ', 'LRW', 'PPT', 'SPN', 'LLA', 'PH*', 'VRD', 'LHR', '*PS', 'DPR', 'ML*', 'LLQ', 'EPM', 'VR*', 'NRS', 'RPV', '*HW', 'LPF', 'QPW', 'TRS', 'CP*', 'NLM', 'SPH', '*LE', 'LHE', 'IHF', 'SLF', 'SHL', 'KPE', 'DP*', 'NHR', 'QLK', 'GPL', 'EH*', '*LG', 'IRL', 'ERI', 'MRK', 'RRT', 'NLG', 'PHT', 'KHM', 'PRY', 'KLS', 'LHD', 'NPY', 'KRI', 'DPD', 'HRK', 'TPR', 'KRT', 'AHW', 'TPT', 'IH*', 'GRA', 'GRG', 'SPM', 'RHM', 'WRK', 'QPR', 'IPE', 'PHD', 'YLH', 'LPL', 'KRC', 'YHI', 'TRT', 'GLL', 'EPK', 'DPL', 'RHK', 'PHI', 'DHP', 'DHS', 'THF', 'WR*', 'WPR', 'SPL', 'PRS', 'KPQ', 'LHT', 'ERR', 'YHL', 'GP*', 'YRP', '*PI', 'WLW', 'PHG', 'GHA', 'WPK', 'MHM', 'DRK', 'DLD', 'HH*', 'VRV', 'IRH', 'AP*', 'YPP', 'CPS', 'LLK', '*PR', 'TLN', 'NHV', 'SP*', 'PLS', 'NHI', 'YPG', 'ALY', 'VLN', 'APQ', 'RPC', 'HPR', 'ARV', 'RLG', 'WPQ', 'KLP', 'MHP', 'VPF', 'RL*', 'HPA', 'EPG', 'THE', 'RLN', 'QLR', 'TRW', 'SRG', 'DRY', 'FPV', 'GLV', 'SRH', 'PPF', 'QRG', 'KHH', 'RPH', 'HLN', 'YPQ', 'GLS', 'SRM', 'ARI', 'KLG', '*PK', 'CRS', 'NPI', 'IPG', 'YPI', 'YRR', 'DHL', 'KRW', 'YPS', 'PLR', 'VLW', 'PHA', 'ILH', 'MRT', 'ALV', 'FRY', 'MPN', 'CPE', 'HPG', 'NHW', 'SHQ', 'EHL', 'VLH', 'CPN', 'PRD', 'RRA', 'YRS', 'ELQ', 'THC', 'DHK', '*RL', 'NLV', 'APF', 'FLV', '*PW', 'THS', 'NPQ', 'THQ', 'VLA', 'YHQ', 'TPH', 'QPL', 'PPS', '*HD', 'EPT', 'FPM', 'QRA', 'KRM', 'RLK', 'LRT', 'QP*', 'MRL', 'DPE', 'FPP', 'AHC', 'EPY', 'RLI', 'ERY', 'CHP', 'DL*', 'KHD', 'MHK', 'WHR', 'PPG', 'YPD', 'ELI', 'ALR', 'IHC', 'PPE', 'QHY', '*LN', 'YLL', '*LD', 'HHP', 'NP*', 'VPM', 'HPK', 'KRQ', 'HPS', 'FHA', 'DRA', 'MLH', 'QRW', 'NH*', 'RLM', 'HPI', 'MRS', 'CHY', 'LLE', 'VLG', 'FPF', 'DHT', 'EHV', 'TRI', 'WLQ', 'VHF', 'QRL', 'SL*', 'ILS', 'GPQ', 'SHY', 'TLM', 'LPK', 'APH', 'KRD', 'FHT', 'QHC', 'KLC', 'NPE', 'DPP', 'PL*', 'NRP', 'YPR', 'AHF', 'GPT', 'QLY', 'VPG', 'TRP', 'ILI', 'PLP', 'VLP', 'GRY', 'HHW', 'QHG', 'CLG', 'RLL', 'SHP', 'IHD', 'ELY', 'QPS', 'KLT', 'ILG', 'YPN', 'DPF', 'RHR', 'HRQ', 'GLM', 'CRI', '*P*', 'AL*', 'DLC', 'VP*', 'AHD', 'ERD', 'SPY', 'FRG', 'KLF', 'MPI', 'MHY', 'ARM', 'SHH', 'HRN', 'ILA', 'TPW', 'KLY', 'ERM', 'WPM', 'TLR', 'CPG', 'MHV', 'WPS', 'SLD', 'HPH', 'RRC', 'RHP', 'VH*', 'DRP', 'LPS', 'GLW', 'ELA', 'GPF', 'FRL', '*RM', 'KHI', 'ELD', 'EHN', '*PA', 'YLQ', 'SRE', 'AHH', 'LH*', 'RRQ', 'RRM', 'VRL', 'MLE', 'YHC', 'RRK', 'KPS', 'VLM', 'IHW', 'MPQ', 'VHP', 'TPN', 'YRD', 'YHN', 'AHG', 'SHF', 'KPG', 'HLH', 'GLI', 'LHV', 'IPT', 'QHF', 'SRL', 'LPV', 'MRN', 'ALA', 'VLI', 'CPD', 'HHF', 'QLC', 'MPP', '*PV', 'LHQ', 'LLT', 'IRK', 'KHT', 'LHM', 'CHH', 'DH*', 'MLV', 'MPE', 'WHQ', 'MPW', 'LLF', 'GHE', 'WRI', 'YP*', 'NRE', 'VPT', 'WHG', 'DRL', 'MPY', 'KPM', 'ERP', 'LLP', 'RHH', 'FRM', 'DLY', 'SRN', 'ELM', 'RPR', 'SRY', 'HLY', 'KRN', 'APE', 'FHL', 'KPP', 'CLI', 'LLN', 'CRM', 'LRL', '*RC', '*LY', 'HLC', 'SLR', 'KPW', 'PPW', 'KPN', 'KRK', 'DHR', 'HHT', 'ARL', '*LK', 'YL*', 'LPN', 'WRM', 'VRS', 'RHE', 'FLL', 'DPC', 'GHL', 'NHP', 'YHG', 'NPV', 'RLP', 'RHT', '*PQ', 'NPD', 'NPL', 'NLR', 'LRC', 'LPD', 'CRT', 'CPK', 'ILK', 'VPY', 'GLF', 'PRG', 'GPV', 'APV', 'IRS', 'QPT', 'ILQ', 'ARN', 'PHY', 'API', 'YLV', 'EPC', 'APK', 'RHF', 'RRW', 'RRP', 'HPQ', 'YLM', 'SLK', 'RP*', 'IPK', 'TRE', 'MRH', 'QLA', 'TPF', 'RHQ', 'QLS', 'AHI', 'MHL', 'NRY', 'LRD', 'MRW', 'TPG', 'KRF', 'CRC', 'HPD', 'IR*', 'FRD', 'NHY', 'HLV', '*HE', 'WPY', 'HRD', 'CHE', 'PRT', 'HRP', 'PLQ', 'WHT', 'WHV', 'HPL', 'MLT', 'NHL', 'SRK', 'FRT', 'PLE', 'LHY', 'TRC', 'EHD', 'RPL', 'TLC', 'ILN', 'DLP', 'HPT', 'RPA', 'GLE', 'VRF', 'KRA', 'TR*', 'ILL', 'DPI', '*RF', 'HRS', 'KHR', 'IRC', 'QRC', 'WPH', 'ALG', 'VPH', 'KPI', 'CLH', 'VHE', 'IRG', 'DHF', 'LLY', 'HRC', 'TRR', '*LH', 'KLD', 'YHW', 'DPS', 'VHH', 'YLP', 'DRT', '*RA', 'GHR', 'RHD', 'WLN', '*RV', 'TPV', 'NHK', 'GLC', 'FRK', 'FLP', 'MPK', 'ERN', 'DLI', 'GPH', 'SLN', 'LHS', 'NHM', 'FRP', 'RHA', 'ARP', 'RRR', 'EPL', 'APR', 'CPM', 'WHS', 'SLE', 'ELL', 'TLV', 'QRQ', 'IHE', 'YH*', 'GRW', '*RT', 'SPI', 'SPF', 'SLH', 'DLA', 'KRV', 'YRG', 'NRA', 'DRS', 'HHQ', 'ELV', 'GLN', 'CLS', 'GR*', 'NRK', 'IPD', 'YRF', 'MHS', 'GLR', 'VRN', 'IRV', 'HLD', '*PC', 'QLM', 'PPK', 'YPT', 'PHK', 'RRF', 'NHT', 'FPY', 'THM', 'WLL', 'QHE', 'SRW', 'IHI', 'QRI', 'PPQ', 'LLH', 'DRM', 'KHN', 'NPH', 'FPC', 'GPE', 'FLA', 'RLY', 'VPC', 'KH*', 'EHK', 'CRR', 'ILY', 'RLS', 'PPH', 'QHD', 'RLW', 'MRM', 'ILE', 'EPA', 'SRP', 'ALM', 'ELR', 'SLA', 'ERV', 'FRC', 'KRL', 'GHH', 'GRM', 'APA', 'HHD', 'MLM', 'LHH', '*HT', 'HHH', 'DRI', 'MPH', 'CPC', '*PY', 'GLH', 'RPE', 'HHK', 'TRA', 'IHV', 'EHG', 'KRS', 'SLM', '*LQ', 'CHV', 'KHY', 'IHN', 'RRE', '*RR', 'FPW', 'MHE'}\n",
      "16383 silent bystander sequences created for test_mutation_1_cftr_alternative\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# define input and output paths and filenames\n",
    "inputpath = './input/'\n",
    "inputfilename = 'input_testfile.csv' # check input_testfile.csv for details about formatting the input file; required columns: [Name, pridict_input, silent, change_edit_bases, in_frame]\n",
    "outputpath = './output/'\n",
    "outputfilename = 'outputfile.csv'\n",
    "#\n",
    "\n",
    "### --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "# default variables (do not change unless you want to adapt the script):\n",
    "silent_surrounding_AA_nr = 2 # number of amino acids up- and downstream of edit AA for which silent bystander mutations will be created; default = 2; maximum 4 AA (up- and downstream) are allowed in this script\n",
    "total_edit_limit = 40 # limit maximum length of edit (including bystander edits) to 40 bases; max. edit length for PRIDICT2.0 predictions is 40 bases\n",
    "max_edit_length = 10 # maximum length of the edit bases you want to change\n",
    "minimum_flanking = 94  # minimum edit-flanking length, after correcting for ORF_start. Only required as sanity check; do not change.\n",
    "### --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "# run bystander input generator\n",
    "inputfiledf, outputfiledf = bystander_input_generator(inputpath, inputfilename, outputpath, outputfilename, silent_surrounding_AA_nr, total_edit_limit, max_edit_length, minimum_flanking)\n",
    "\n",
    "# continue with running PRIDICT2 (outside of this notebook) with the outputfile.csv as input file\n",
    "# example command: python pridict2_pegRNA_design.py batch --input-dir ./silentbystander_addon/output --input-fname outputfile.csv --output-dir ./predictions\n",
    "\n",
    "# Optional but NOT RECOMMENDED: run PRIDICT2 from within this notebook. (uncomment !python command below)\n",
    "# Caveat: takes a LONG time to run; we recommend running it separately via commandline\n",
    "# !python ../pridict2_pegRNA_design.py batch --input-dir ./output --input-fname outputfile.csv --output-dir ../predictions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summarize PRIDICT2.0 predictions of silent bystanders after running PRIDICT2.0\n",
    "- Only run this after you ran PRIDICT2.0 with the output batch file created above. \n",
    "\n",
    "- The code below summarizes all the predictions with different silent bystanders in one file, sorts this by K562 score and saves it as summary prediction file.\n",
    "\n",
    "- For MMR-deficient context, change \"sort_value\" from \"K562\" to \"HEK\".\n",
    "\n",
    "- From this summary file, we suggest to take e.g. the top 5 pegRNAs and test these in your experimental setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Summarize PRIDICT2 predictions:\n",
    "pridict2_predictions_folder = '../predictions/'  # folder with all PRIDICT2 predictions (default \"../predictions/\")\n",
    "summary_prediction_output_folder = './summarized_silent_pridict2_predictions/' # folder where summarized prediction files will be saved\n",
    "sort_value = 'K562' # change to \"HEK\" for sorting to MMR-deficient cell line prediction\n",
    "\n",
    "# filelist of all .csv files in pridict2_predictions_folder:\n",
    "filelist = [f for f in os.listdir(pridict2_predictions_folder) if f.endswith('pegRNA_Pridict_full.csv')]\n",
    "\n",
    "for index, row in inputfiledf.iterrows():\n",
    "    sequence_name = row['Name']\n",
    "    # get all files in filelist that start with the sequence_name\n",
    "    sequence_files = [f for f in filelist if f.startswith(sequence_name)]\n",
    "    # read all files and concatenate them\n",
    "    all_files = []\n",
    "    for file in sequence_files:\n",
    "        all_files.append(pd.read_csv(pridict2_predictions_folder+file))\n",
    "    all_files_df = pd.concat(all_files, ignore_index=True)\n",
    "    # sort all_files_df by column \"PRIDICT2_0_editing_Score_deep_...\" (from highest to lowest)\n",
    "    all_files_df = all_files_df.sort_values(by='PRIDICT2_0_editing_Score_deep_'+sort_value, ascending=False)\n",
    "    # save concatenated file\n",
    "    all_files_df.to_csv(summary_prediction_output_folder+sequence_name+'_all_silent_predictions.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "spyderenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
